### PR TITLE
Core: Improve token exchange handling when token expires

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 1]
```diff
@@ -159,22 +158,26 @@ public void initialize(String name, Map<String, String> unresolved) {
...
     this.sessions = newSessionCache(mergedProps);
     this.refreshAuthByDefault =
         PropertyUtil.propertyAsBoolean(
             mergedProps,
             CatalogProperties.AUTH_DEFAULT_REFRESH_ENABLED,
             CatalogProperties.AUTH_DEFAULT_REFRESH_ENABLED_DEFAULT);
-    this.client = clientBuilder.apply(mergedProps);
     this.paths = ResourcePaths.forCatalogProperties(mergedProps);
 
+    this.catalogAuth = new AuthSession(baseHeaders, null, null);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:170]: these are mainly all refactorings that have been already done in https://github.com/apache/iceberg/pull/6169. This is so that the token refresh mechanism can be used both for the `RESTSessionCatalog` and the Signer client

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 2]
```diff
@@ -787,9 +762,39 @@ private static ConfigResponse fetchConfig(
   private AuthSession newSession(
       Map<String, String> credentials, Map<String, String> properties, AuthSession parent) {
     if (credentials != null) {
-      // use the bearer token without exchanging
       if (credentials.containsKey(OAuth2Properties.TOKEN)) {
-        return newSession(credentials.get(OAuth2Properties.TOKEN), expiresInMs(properties), parent);
+        String token = credentials.get(OAuth2Properties.TOKEN);
+        Optional<JwtToken> jwt = JwtToken.of(token);
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:767]: here's the main logic that was changed

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 3]
```diff
@@ -376,5 +409,105 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+                  refreshStartTime,
+                  expiration.first(),
+                  expiration.second());
+            }
+          },
+          timeToWait,
+          TimeUnit.MILLISECONDS);
+    }
+
+    public static AuthSession sessionFromToken(
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:450]: those 3 methods are mainly public so that they can also be used for https://github.com/apache/iceberg/pull/6169

File: core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  default long expiresInMillis() {
+    return expiresAtEpochMillis() - Instant.now().toEpochMilli();
+  }
+
+  default boolean expiredAt(long expiredAtEpochMillis) {
+    return expiresAtEpochMillis() < expiredAtEpochMillis;
+  }
+
+  static Optional<JwtToken> of(String token) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java:44]: given that we can't assume that we have a JWT token, it's probably best to return here an `Optional<JwtToken>` and only check expiration on the JWT token if we have one

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 5]
```diff
@@ -159,22 +158,26 @@ public void initialize(String name, Map<String, String> unresolved) {
...
-    this.client = clientBuilder.apply(mergedProps);
     this.paths = ResourcePaths.forCatalogProperties(mergedProps);
 
+    this.catalogAuth = new AuthSession(baseHeaders, null, null);
+    if (authResponse != null) {
+      this.catalogAuth =
+          AuthSession.sessionFromFetchedToken(
+              client, tokenRefreshExecutor(), authResponse, startTimeMillis, catalogAuth);
+    } else if (initToken != null) {
+      this.catalogAuth =
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:176]: I'm currently not 100% sure whether we already should check here if the `initToken` is expired and do a token exchange using the credential (similarly to how it's down further down in this class)

File: core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java (hunk 4 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java:44]: makes perfect sense, updated :)

File: core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    try {
+      Preconditions.checkArgument(null != token, "Invalid Bearer token: null");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/JwtToken.java:46]: I don't think it makes sense to use "Bearer" here. This is possibly a JWT and those are sometimes used for bearer authorization, but not necessarily. No need to mix the two concepts.

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Optional;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+import org.immutables.value.Value;
+
+@Value.Immutable
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:29]: I think that `@Immutable` is being overused for cases like this. This class isn't complex and using it forces a lot of logic to go into default methods and  the `of` static method. While it is useful in some cases, I think that there should be a significant savings from using immutables; we should not use them by default.

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  String token();
+
+  long expiresAtEpochMillis();
+
+  default long expiresInMillis() {
+    return expiresAtEpochMillis() - Instant.now().toEpochMilli();
+  }
+
+  default boolean expiredAt(long expiredAtEpochMillis) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:40]: We do usually follow the convention that methods that return a boolean should start with `is` or `has` or some other word that works with an `if` or `while` context.

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+  static Optional<JWT> of(String token) {
+    try {
+      Preconditions.checkArgument(null != token, "Invalid JWT: null");
+      String[] parts = token.split("\\.");
+      Preconditions.checkArgument(parts.length == 3, "Invalid JWT: %s", token);
+
+      // Parse the payload JSON
+      JsonNode jsonNode = JsonUtil.mapper().readTree(Base64.getUrlDecoder().decode(parts[1]));
+      Long expiresAt = JsonUtil.getLongOrNull("exp", jsonNode);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:52]: Do we need a class for JWT, or could this be a util method in `OAuth2Util`?

If it were a method in the util class, then we could support more token types or patterns (not just `exp`) in the future.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 10]
```diff
@@ -101,6 +106,17 @@ public static Map<String, String> authHeaders(String token) {
     }
   }
 
+  public static Map<String, String> authBasicHeader(String credential) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:109]: Shouldn't it be `basicAuth` rather than `authBasic`? The other methods that return a map also use a plural, `Headers` because there could be more than one.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 11]
```diff
@@ -302,8 +318,16 @@ public static OAuthTokenResponse tokenResponseFromJson(JsonNode json) {
     return builder.build();
   }
 
+  public static boolean tokenExpired(Optional<JWT> jwt) {
+    return jwt.isPresent()
+        && (jwt.get()
+            .expiredAt(System.currentTimeMillis() + AuthSession.MAX_REFRESH_WINDOW_MILLIS));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:324]: I think that checking whether a token is expired should check the current time, not whether it is expected to expire in 5 minutes.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 5 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:176]: Yes, this should check whether the expiration timestamp can be recovered and schedule based on that.

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java (hunk 9 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:52]: I initially had this inside a method but the downside was basically that you'd need to parse the token multiple times if you'd wanted to access token expiration multiple times

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 12]
```diff
@@ -376,5 +409,105 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+          },
+          timeToWait,
+          TimeUnit.MILLISECONDS);
+    }
+
+    public static AuthSession sessionFromToken(
+        RESTClient client,
+        ScheduledExecutorService tokenRefreshExecutor,
+        String token,
+        Long expiresIn,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:454]: Now that this is public, this duration needs to have a specific unit, like `expiresInMillis`.

I also think that this should be `defaultExpirationMillis` and that the token should be parsed in this method to get its specific expiration, if possible. That would avoid leaking the token parsing logic outside of this class, I think.

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java [hunk 13]
```diff
@@ -787,9 +762,39 @@ private static ConfigResponse fetchConfig(
   private AuthSession newSession(
       Map<String, String> credentials, Map<String, String> properties, AuthSession parent) {
     if (credentials != null) {
-      // use the bearer token without exchanging
       if (credentials.containsKey(OAuth2Properties.TOKEN)) {
-        return newSession(credentials.get(OAuth2Properties.TOKEN), expiresInMs(properties), parent);
+        String token = credentials.get(OAuth2Properties.TOKEN);
+        Optional<JWT> jwt = JWT.of(token);
+
+        if (OAuth2Util.tokenExpired(jwt)) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:769]: The logic for token expiration should be handled in `sessionFromToken` and other OAuth2Util methods.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 14]
```diff
@@ -376,5 +409,105 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+      long waitIntervalMillis = expiresInMillis - refreshWindowMillis;
+      // how much time has already elapsed since the new token was issued
+      long elapsedMillis = System.currentTimeMillis() - startTimeMillis;
+      // how much time to actually wait
+      long timeToWait = Math.max(waitIntervalMillis - elapsedMillis, MIN_REFRESH_WAIT_MILLIS);
+
+      tokenRefreshExecutor.schedule(
+          () -> {
+            long refreshStartTime = System.currentTimeMillis();
+            Pair<Integer, TimeUnit> expiration = session.refresh(client);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:435]: The refresh done here also needs to handle cases where the token has expired.

**Review by PraveenNanda124 — COMMENTED** [review]
Looks good

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 11 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:324]: makes sense, updated

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 12 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:454]: this makes sense. I've updated the code accordingly

File: core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java (hunk 5 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:176]: moving the token expiration logic to `sessionFromToken` solves this

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 14 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:430]: what specifically did you have in mind here? I looked at the logic in https://github.com/apache/iceberg/blob/3f01ed874bea66a05195e491c9b7045a4337d902/core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java#L368 and it seems to do the right thing (as it relies on the expiration time that is being part of the response)

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 15]
```diff
@@ -376,5 +404,127 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+        RESTClient client,
+        ScheduledExecutorService tokenRefreshExecutor,
+        String token,
+        @Nullable String credential,
+        Long defaultExpirationMillis,
+        AuthSession parent) {
+      Optional<JWT> jwt = JWT.of(token);
+
+      if (jwt.isPresent() && jwt.get().isExpired()) {
+        Preconditions.checkState(
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:455]: I think it is better to be explicit here and fail rather than silently adding empty headers if the `credential` is `null`

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 16]
```diff
@@ -101,6 +107,17 @@ public static Map<String, String> authHeaders(String token) {
     }
   }
 
+  public static Map<String, String> basicAuthHeader(String credential) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:110]: I think this still needs to be plural.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 14 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:430]: I'm referring to retries. If the refresh fails because the token has expired, then this should retry using basic auth, if possible.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 17]
```diff
@@ -376,5 +404,127 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+            client,
+            tokenRefreshExecutor,
+            token,
+            OAuth2Properties.ACCESS_TOKEN_TYPE,
+            session,
+            OAuth2Properties.CATALOG_SCOPE);
+      } else {
+        AuthSession session =
+            new AuthSession(parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE);
+        Long expiresInMillis = jwt.map(JWT::expiresInMillis).orElse(defaultExpirationMillis);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:472]: Can this be a primitive?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 18]
```diff
@@ -376,5 +404,127 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+          },
+          timeToWait,
+          TimeUnit.MILLISECONDS);
+    }
+
+    public static AuthSession sessionFromToken(
+        RESTClient client,
+        ScheduledExecutorService tokenRefreshExecutor,
+        String token,
+        @Nullable String credential,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:449]: Iceberg doesn't use `@Nullable` annotations. What's the value of it here?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 19]
```diff
@@ -376,5 +404,127 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+        Long defaultExpirationMillis,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:450]: Can this be a primitive or are there cases where it isn't known?

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java (hunk 9 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:52]: Those are both in the same method. Why not just use a local variable?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 20]
```diff
@@ -376,5 +404,127 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+            new AuthSession(parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE);
+        Long expiresInMillis = jwt.map(JWT::expiresInMillis).orElse(defaultExpirationMillis);
+
+        if (null != expiresInMillis) {
+          scheduleTokenRefresh(
+              client,
+              tokenRefreshExecutor,
+              session,
+              System.currentTimeMillis(),
+              expiresInMillis,
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:480]: I think this should pass in the expiration time directly, rather than converting to an interval and passing that. The existing refresh method is based on `expiresIn` because that's what we get back from the request, but if we have a known timestamp instead it is more accurate to use it.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:480]: This is probably better to fix in a follow up... let's leave it as-is for now and we can convert everything to `expiresAt` later.

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java (hunk 7 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:29]: I think this should be rewritten to remove `@Immutable`.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 18 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:449]: this was meant for Dev documentation purposes to indicate that the `credential` can in fact be null in some cases

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 17 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:472]: `defaultExpirationMillis` could still be null due to `private Long expiresInMs(Map<String, String> properties)` potentially returning a null, so I think it needs to be a `Long` rather than a primitive

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 16 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:110]: I've adjusted this to be plural to be in line with `authHeaders` but it feels weird that those are plural, since technically there's always only a single authorization header being added to the map

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 19 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:450]: this could still be null because `private Long expiresInMs(Map<String, String> properties)` might return null

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java (hunk 7 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:29]: I understand your argument about potentially over-using `@Immutable`. However, I think there's rarely a case where the non-`@Immutable` version is better than the `@Immutable` one. In fact I initially put the JWT parsing code first into a method, then extracted it into a non-`@Immutable` class like the one below and eventually made it `@Immutable`. 
The non-immutable version of that class will be longer due to boiler-plate code. Technically the constructor might even have a precondition null check for the token (which we don't need in this particular example, because we already know the token is non-null from the static method). In the `@Immutable` we get this stuff for free because the API definition is explicit (aka the `token` isn't marked as optional or nullable and therefore must always be non-null).

```
public class NonImmutableJwt {

  private final String token;
  private final long expiresAtEpochMillis;

  private Jwt(String token, long expiresAtEpochMillis) {
    this.token = token;
    this.expiresAtEpochMillis = expiresAtEpochMillis;
  }

  public String token() {
    return token;
  }

  public long expiresAtEpochMillis() {
    return expiresAtEpochMillis;
  }

  public long expiresInMillis() {
    return expiresAtEpochMillis() - Instant.now().toEpochMilli();
  }

  public boolean isExpired() {
    return expiresAtEpochMillis() <= Instant.now().toEpochMilli();
  }

  static Optional<Jwt> of(String token) {
    try {
      Preconditions.checkArgument(null != token, "Invalid JWT: null");
      String[] parts = token.split("\\.");
      Preconditions.checkArgument(parts.length == 3, "Invalid JWT: %s", token);

      // Parse the payload JSON
      JsonNode jsonNode = JsonUtil.mapper().readTree(Base64.getUrlDecoder().decode(parts[1]));
      Long expiresAt = JsonUtil.getLongOrNull("exp", jsonNode);
      return Optional.of(
          new Jwt(
              token,
              null == expiresAt
                  ? Long.MAX_VALUE
                  : Instant.ofEpochSecond(expiresAt).toEpochMilli()));
    } catch (Exception e) {
      return Optional.empty();
    }
  }
}
```
If you're strong on this one, then I can change this to a non-`@Immutable` version but as I've tried to outline above, I don't think there's any advantage in doing so.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 14 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:448]: I've pushed something where we try and refresh the token using the credential when the normal retries fail. Let me know if this is what you had in mind

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 21]
```diff
@@ -344,7 +382,20 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
             Tasks.foreach(ref)
                 .suppressFailureWhenFinished()
                 .retry(5)
-                .onFailure((task, err) -> LOG.warn("Failed to refresh token", err))
+                .onFailure(
+                    (holder, err) -> {
+                      if (null != credential) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:387]: I have a test for this behavior but it needs some additional changes to the code (parameterizing the number of retries that are being done so that the default of 5 can be lowered to 1 for testing). @rdblue are we ok adding that test + the code changes for parameterizing the number of retries as a separate PR?

File: core/src/main/java/org/apache/iceberg/rest/auth/JWT.java (hunk 9 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/JWT.java:52]: sorry I'm not sure I'm following/understanding your argument properly. Could you elaborate here please what kind of change you'd like to see here?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 20 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:505]: I've refactored this part of the code in a separate PR (https://github.com/apache/iceberg/pull/6562/commits/7c4655ea6d5390301106c2f5f2d51dfc63adf779)

**Comment by rdblue** [comment]
@nastra, I opened a PR against your branch with the remaining changes I think we should make. Please take a look: https://github.com/nastra/iceberg/pull/72

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 21 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:387]: Yes, a separate PR is fine.

**Comment by nastra** [comment]
> @nastra, I opened a PR against your branch with the remaining changes I think we should make. Please take a look: [nastra#72](https://github.com/nastra/iceberg/pull/72)

Thanks @rdblue I've applied your changes and cleaned up some tiny bits to make CI pass. I've also made sure that the changes here are compatible for https://github.com/apache/iceberg/pull/6169.

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 22]
```diff
@@ -302,17 +317,59 @@ public static OAuthTokenResponse tokenResponseFromJson(JsonNode json) {
...
+    private static final long MIN_REFRESH_WAIT_MILLIS = 10;
     private Map<String, String> headers;
     private String token;
     private String tokenType;
+    private Long expiresAtMillis;
+    private final String credential;
     private volatile boolean keepRefreshed = true;
 
     public AuthSession(Map<String, String> baseHeaders, String token, String tokenType) {
+      this(baseHeaders, token, tokenType, null);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:363]: Should we deprecate this constructor so we can remove it in 1.3.0?

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java [hunk 23]
```diff
@@ -376,5 +450,201 @@ public Pair<Integer, TimeUnit> refresh(RESTClient client) {
...
+      if (isExpired(expiresAtMillis, System.currentTimeMillis())) {
+        // the token has already expired, attempt to refresh using the credential
+        return refreshExpiredToken(client);
+      } else {
+        // attempt a normal refresh
+        return refreshToken(client, headers(), token, tokenType, OAuth2Properties.CATALOG_SCOPE);
+      }
+    }
+
+    static boolean isExpired(Long expiresAtMillis, long now) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:464]: The reason why I didn't add a method like this is that it isn't correct when `expiresAtMillis` is `null`. It returns false, but the token may still be expired. I think it's misleading to have a method with this name, so I think we should revert the change to add this.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
I think that we need to fix https://github.com/apache/iceberg/pull/6489/files#r1072451113 in a follow-up, but that's minor so I'm going to merge this to unblock the next steps. Thanks, @nastra!

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 23 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:464]: thanks for bringing this up. I'll look into that as part of https://github.com/apache/iceberg/pull/6562

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 22 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java:363]: good point, will do that

File: core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java (hunk 20 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java]: here's the PR that refactors this: https://github.com/apache/iceberg/pull/6562
